algopy_testing/src/algopy_testing/models/global_state.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, TypedDict, TypeVar
+
+if TYPE_CHECKING:
+    from algopy_testing.models.account import Account
+    from algopy_testing.models.application import Application
+    from algopy_testing.primitives.bytes import Bytes
+    from algopy_testing.primitives.uint64 import UInt64
+
+T = TypeVar("T")
+
+
+class GlobalFieldsKwargs(TypedDict, total=False):
```

Comment:
I think we should avoid having two ways of defining these values, tried this approach with algokit_utils but it ended up just creating more issues, so would want to avoid doing that again. My preference would be to just have the dataclass style (with default values) as this gives better IDE support and the only con is it's slightly more verbose

Reply:
@daniel-makerx the problem with a single dataclass per abstraction in models (for things like global, txn, itxn) was the lack of straightforward way to allow users to have typed individual fields for each abstraction that they can set on the context via methods like patch_global_fields and etc. 

@daniel-makerx let me know what you think of the latest implementation. 

I am using a TypedDict (GlobalFields and etc) per each and then dynamically fetch the values in __getattr__ dunder in the dataclass, the exported object is an instance of the dataclass (given that the __getattr__ only gets invoked when accessing attributes on instances). 

As a last tweak, i am also using Unpack from `typing` to unpack the TypedDict as a **kwargs field a user can pass on the add_* methods. That way a user gets to be able to easily patch any field on the Global, Txn and ITxn - and then within mocked contract execution accessing the values like Global will in turn access an instance of a dataclass where the getattr dunder will take care of loading the patched field (if exists) from the dictionary set on the global context instance. 

I believe this still addresses the maintainability issue of having the same field duplicated on typeddict vs dataclass